src/buf_writer.c

```diff
@@ -5,8 +5,43 @@
 #include "jemalloc/internal/buf_writer.h"
 #include "jemalloc/internal/malloc_io.h"
 
+static void *
+buf_writer_allocate_internal_buf(size_t buf_len) {
+	if (buf_len > SC_LARGE_MAXCLASS) {
+		return NULL;
+	}
+	return iallocztm(TSDN_NULL, buf_len, sz_size2index(buf_len), false,
+	    NULL, true, arena_get(TSDN_NULL, 0, true), true);
+}
+
+static void
+buf_writer_free_internal_buf(void *buf) {
+	idalloctm(TSDN_NULL, buf, NULL, NULL, true, true);
+}
+
+bool
+buf_writer_init(buf_writer_t *buf_writer, write_cb_t *write_cb, void *cbopaque,
+    char *buf, size_t buf_len) {
+	buf_writer->write_cb = write_cb;
+	buf_writer->cbopaque = cbopaque;
+	assert(buf_len >= 2);
+	if (buf != NULL) {
+		buf_writer->buf = buf;
+		buf_writer->internal_buf = false;
+	} else {
+		buf_writer->buf = buf_writer_allocate_internal_buf(buf_len);
+		buf_writer->internal_buf = true;
+	}
+	buf_writer->buf_size = buf_len - 1; /* Allowing for '\0' at the end. */
+	buf_writer->buf_end = 0;
+	return buf_writer->buf == NULL;
```

Comment:
nit: if allocation failed, better return early instead of set up everything. Maybe alloc at the beginning.

Reply:
It looks the intention is to make best effort and keep going. In that case, at least make the buf size 0.

Reply:
Makes sense. Also added unit test to verify.